libp2p/switch.nim

```diff
@@ -281,45 +320,9 @@ proc internalConnect(s: Switch,
         raise newException(DialFailedError, "Zombie connection encountered")
 
       trace "Reusing existing connection", conn, direction = $conn.dir
-
       return conn
 
-    debug "Dialing peer", peerId
-    for t in s.transports: # for each transport
-      for a in addrs: # for each address
-        if t.handles(a):   # check if it can dial it
-          trace "Dialing address", address = $a, peerId
-          let dialed = try:
-              await t.dial(a)
-            except CancelledError as exc:
-              debug "Dialing canceled", msg = exc.msg, peerId
-              raise exc
-            except CatchableError as exc:
-              debug "Dialing failed", msg = exc.msg, peerId
-              libp2p_failed_dials.inc()
-              continue # Try the next address
-
-          # make sure to assign the peer to the connection
-          dialed.peerInfo = PeerInfo.init(peerId, addrs)
-
-          libp2p_dialed_peers.inc()
-
-          let upgraded = try:
-              await s.upgradeOutgoing(dialed)
-            except CatchableError as exc:
-              # If we failed to establish the connection through one transport,
-              # we won't succeeded through another - no use in trying again
-              await dialed.close()
-              debug "Upgrade failed", msg = exc.msg, peerId
-              if exc isnot CancelledError:
-                libp2p_failed_upgrade.inc()
-              raise exc
-
-          doAssert not isNil(upgraded), "connection died after upgradeOutgoing"
-
-          conn = upgraded
-          debug "Dial successful", conn, peerInfo = conn.peerInfo
-          break
+    conn = await s.dialAndUpgrade(peerId, addrs)
   finally:
     if lock.locked():
```

Comment:
so this if shouldn't really be here?

Reply:
or maybe it should.. if acquire is cancelled, I guess.. another way to see it is that acquire should not be in the try/catch

Reply:
If `acquire` isn't in the try/catch, then you run the risk of never releasing.

Reply:
that is for a well functioning try/catch.

Reply:
acquire doesn't raise (except cancellation) so it can be outside, but it doesn't really matter - the if is needed when the acquire is in the try/catch because on cancellation, it will not be acquired